### PR TITLE
Don't allow network address and broadcast address

### PR DESCRIPTION
**Pull request by yaocw2020**
To fix the case like
image

Related issue: https://github.com/harvester/harvester/issues/1458

Test plan:
image

image

image

### CODE REVIEW

File: pkg/ipam/allocator.go [hunk 1]
```diff
@@ -135,6 +112,25 @@ func MakeRange(r *lbv1.Range) (*allocator.Range, error) {
...
+	}
+
+	ip := net.ParseIP(ipStr)
+	if ip == nil {
+		return nil, fmt.Errorf("invalid IP %s", ipStr)
+	}
+	if !ipNet.Contains(ip) {
+		return nil, fmt.Errorf("IP %s is out of subnet %s", ipStr, ipNet.String())
+	}
+	if ip.Equal(networkIP(*ipNet)) || ip.Equal(broadcastIP(*ipNet)) {
```
Comment (futuretea) [inline pkg/ipam/allocator.go:127]: Separate the two types of errors

File: pkg/ipam/allocator.go [hunk 2]
```diff
@@ -66,42 +66,35 @@ func MakeRange(r *lbv1.Range) (*allocator.Range, error) {
...
 	} else {
-		start = net.ParseIP(r.RangeStart)
-		if start == nil {
-			return nil, fmt.Errorf("invalid rangeStart %s", r.RangeStart)
-		}
-
-		if !ipNet.Contains(start) {
-			return nil, fmt.Errorf("range start IP %s is out of subnet %s", start.String(), ipNet.String())
-		}
+		// The rangeStart defaults to “.1” IP inside the “subnet” block.
```
Comment (futuretea) [inline pkg/ipam/allocator.go:77]: replace “” to "" or ``

File: pkg/ipam/allocator.go [hunk 3]
```diff
@@ -135,6 +112,25 @@ func MakeRange(r *lbv1.Range) (*allocator.Range, error) {
 	}, nil
 }
 
+func parseIP(ipStr string, ipNet *net.IPNet, defaultIP net.IP) (net.IP, error) {
+	if ipStr == "" {
+		return defaultIP, nil
+	}
+
+	ip := net.ParseIP(ipStr)
```
Comment (chrisho) [inline pkg/ipam/allocator.go:120]: The variable name `ip` is conflict with the package `github.com/containernetworking/plugins/pkg/ip`

File: pkg/ipam/allocator.go [hunk 4]
```diff
@@ -66,42 +66,35 @@ func MakeRange(r *lbv1.Range) (*allocator.Range, error) {
...
 	mask := ipNet.Mask.String()
-	// The rangeStart defaults to “.1” IP inside the “subnet” block.
-	if r.RangeStart == "" {
-		// To return the IP with 16 bytes representation as same as what the function net.ParseIP returns
-		if mask == p2pMaskStr {
-			start = ipv4.To16()
-		} else {
-			start = ip.NextIP(ipNet.IP).To16()
-		}
+	// If the subnet is point to point IP
```
[the remark this answers is not in the document]
Comment (yaocw2020) [reply pkg/ipam/allocator.go]: It's "github.com/containernetworking/plugins/pkg/ip". I have given it an alias `cnip`.

File: pkg/ipam/allocator.go (hunk 3 above)
Comment (yaocw2020) [reply pkg/ipam/allocator.go:125]: I have given it an alias `cnip`.

**Review by chrisho — APPROVED** [review]
LGTM and the validator work fine, thanks.

**Review by guangbochen — APPROVED** [review]

**Review by futuretea — APPROVED** [review]
LGTM, thanks.

**Pull request merged**